### PR TITLE
vp-604_show_inherited_tags

### PR DESCRIPTION
**Pull request by AliveMen**
show tags total count with inherited tags
show inherited tags

tags

### CODE REVIEW

File: VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs [hunk 1]
```diff
@@ -71,9 +106,16 @@ public IHttpActionResult GetTagsCount(string id)
...
             };
+
             var result = _searchService.SearchTaggedItems(criteria).Results.FirstOrDefault();
             var count = result?.Tags.Count ?? 0;
 
+            var entities = _taggedEntitiesServiceFactory.Create(KnownDocumentTypes.Category).GetEntitiesByIds(new[] { id });
+            if (!entities.IsNullOrEmpty())
+            {
+                var effectiveTagsDict = _tagPropagationPolicy.GetResultingTags(entities);
+                count = effectiveTagsDict[id].Count;
```
Comment (yecli) [inline VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: We need to count unique tags, as own and inherited ones could be duplicated.

Comment (AliveMen) [reply VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: fixed

File: VirtoCommerce.CatalogPersonalizationModule.Core/Model/EffectiveTag.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+namespace VirtoCommerce.CatalogPersonalizationModule.Core.Model
+{
+    public class EffectiveTag
```
Comment (tatarincev) [inline VirtoCommerce.CatalogPersonalizationModule.Core/Model/EffectiveTag.cs]: Please make a two static fabric  methods
EffectiveTag  InheritedTag(tag)
EffectiveTag NonInheritedTag(tag)

and use them for the instances creation

Comment (AliveMen) [reply VirtoCommerce.CatalogPersonalizationModule.Core/Model/EffectiveTag.cs]: done

File: VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs [hunk 3]
```diff
@@ -69,10 +71,12 @@ public IHttpActionResult GetTagsCount(string id)
...
             {
                 EntityId = id,
-                Take = 1
+                Take = 1,
+                ResponseGroup = TaggedItemResponseGroup.WithInheritedTags.ToString()
             };
+
             var result = _searchService.SearchTaggedItems(criteria).Results.FirstOrDefault();
-            var count = result?.Tags.Count ?? 0;
+            var count = result?.Tags.Union(result.InheritedTags).Distinct().Count();
```
Comment (yecli) [inline VirtoCommerce.CatalogPersonalizationModule.Web/Controllers/Api/PersonalizationModuleController.cs]: Probably need to count Null possibility:
`result.Tags.Union(result?.InheritedTags).Distinct().Count() ?? 0`

File: VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs [hunk 4]
```diff
@@ -7,117 +8,194 @@
...
+                evaluatedEntityIds = result.Select(x => x.EntityId).ToList();
+                foreach (var taggedItem in result)
+                {
+                    var entities = _taggedEntitiesServiceFactory.Create(taggedItem.EntityType).GetEntitiesByIds(new[] { taggedItem.EntityId }).ToList();
+                    var evaluatedItems = EvaluateEffectiveTags(entities);
+                    taggedItem.InheritedTags = evaluatedItems.FirstOrDefault(x => x.EntityId == taggedItem.EntityId)?.InheritedTags;
+                }
+            }
+
+            entityIds = entityIds.Where(ids => evaluatedEntityIds.All(evaluatedIds => evaluatedIds != ids)).ToArray();
```
Comment (yecli) [inline VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs]: What this code is doing? Seems it is not get all ids except evaluatedIds.
I think there is linq [Except](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.except?view=netframework-4.8) function for this.

File: VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs [hunk 5]
```diff
@@ -7,117 +8,198 @@
...
+                }
+                query = query.OrderBySortInfos(sortInfos);
+
+                retVal.TotalCount = query.Count();
+                query = query.Skip(criteria.Skip).Take(criteria.Take);
+
+                var ids = query.Select(x => x.Id).ToArray();
+                retVal.Results = GetTaggedItemsByIds(ids, criteria.ResponseGroup).AsQueryable().OrderBySortInfos(sortInfos).ToList();
+
+                if (criteria.ResponseGroup.Contains(TaggedItemResponseGroup.WithInheritedTags.ToString()) &&
```
Comment (tatarincev) [inline VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs]: It is the really weird way to check flags, why you do this by this way?  We have a lot of example how to do this? could you check? https://github.com/VirtoCommerce/vc-module-catalog/blob/dev/VirtoCommerce.CatalogModule.Data/Repositories/CatalogRepositoryImpl.cs#L275

**Comment by yecli** [comment]
https://github.com/VirtoCommerce/vc-module-catalog-personalization/releases/tag/v1.1.4

**Pull request merged**